### PR TITLE
feat(release workflows): Improve Release workflows to enable bugfix versions

### LINKED ISSUE #1448: Release workflow: make base branch configurable
improvement to the release flow: 
  - configurable base: currently releases are always created off of `main`, but for isolated bugfixes we will need an option to do it off of some other branch
  - duplicated branches: we currently have `releases` and `release/x.y.z`, which seems spurious. We can probably skip the `releases` branch altogether.

### PR DESCRIPTION
**Pull request by rafaelmag110**
## WHAT

Implements the decisions made in the following DR: https://github.com/eclipse-tractusx/tractusx-edc/tree/main/docs/development/decision-records/2024-08-28-Release-Workflow-Refactor

- The PR that triggers the release now has the `main` branch as base instead of `releases`. No need to be moving commits around from releases and then to main.
- A new action `publish-tag-and-release` was created to enable a parametrized execution of the previous GH tag and release set of steps, depending if its a latest or bugfix release.
- Release version is now resolved as the version in `gradle.properties` and not from branch names.
- A manual workflow to trigger manual bugfix releases was introduced. It basically re-uses the release workflow.
- Adds a validation step to skip the run if the tag to be created already exists.
- Adds a gate that prevents unintentional workflow runs in forks.

## WHY

- Enable bugfix versions.
- Refactor existing release workflows to remove the need for a `releases/ ` branch and even `release/*` branches.
- General CI improvements.

Closes #1448

### CODE REVIEW

**Comment by rafaelmag110** [comment]
@paullatzelsperger 
@wolf4ood 
Please review.

File: .github/workflows/draft-release.yaml
Comment (paullatzelsperger) [inline .github/workflows/draft-release.yaml:35]: nit: I would put a more explicit description like "Validates that tag <TAG> does not already exist"

File: .github/workflows/draft-release.yaml
Comment (paullatzelsperger) [inline .github/workflows/draft-release.yaml:61]: Personally, I prefer a simple shell script over conditional steps or dedicated actions, because shell scripts can be easily replicated in a local environment and are generally more easily understandable.

**Review by paullatzelsperger — COMMENTED** [review]
minor remarks, looks good to me otherwise

File: .github/actions/publish-tag-and-release/action.yml
Comment (ndr-brt) [inline .github/actions/publish-tag-and-release/action.yml:32]: nit: in other composite actions we used different case. the most used looks to be the snake_case. 
(for the record, in yaml files for some reasons I like kebab-case more, but snake will fit better in this project)

File: .github/actions/publish-tag-and-release/action.yml
Comment (ndr-brt) [inline .github/actions/publish-tag-and-release/action.yml:55]: using `--force` does not look really safe

File: .github/actions/publish-tag-and-release/action.yml
Comment (ndr-brt) [inline .github/actions/publish-tag-and-release/action.yml:82]: I'd say that new snapshot on main branch should be:
`"$RELEASE_VERSION_MAJOR.$((RELEASE_VERSION_MINOR+$INC)).0"-SNAPSHOT`

File: .github/workflows/release.yml
Comment (ndr-brt) [inline .github/workflows/release.yml:40]: this looks complicated plus it will be evaluated for every PR closed. given that we're following a sort-of gitflow approach, why don't we release on the `release` branch and at the end we merge it back to `main` later? this will simplify things a little.

File: .github/actions/publish-tag-and-release/action.yml
Comment (rafaelmag110) [reply .github/actions/publish-tag-and-release/action.yml:55]: Agree. I overlooked this and figured it was needed at some point. Maybe to override existing tags? I don't think it's a good practice anyway.

File: .github/workflows/release.yml
Comment (rafaelmag110) [reply .github/workflows/release.yml:40]: Yes, it will be evaluated for every merged PR on `main` but it won't run because of the condition. And since this is a job-level conditional, github will just skip its execution and not throw any error (btw, that is the reason why I prefer job/action level conditionals vs everything script).

Like such.

image


If we release from the `release` branch we can't create the tag from it, otherwise the release branch cannot be deleted.

I wouldn't say it looks complicated. Its actually very readable. It's just a veeeery long line 😅

Comment (ndr-brt) [reply .github/workflows/release.yml:40]: well, the tag can be created after the release branch has been merged into main. I think that's the typical gitflow approach

File: .github/actions/publish-tag-and-release/action.yml
Comment (ndr-brt) [reply .github/actions/publish-tag-and-release/action.yml:55]: I would just remove that, the only advantage is that it speeds-up the process in case of re-launch (because of a failure) but it exposes the release to affect older releases.

File: .github/workflows/release.yml
Comment (rafaelmag110) [reply .github/workflows/release.yml:40]: I don't think I got your point then, sorry.
Your suggestion is to change the trigger of the release workflow to be the push on the `release` branch?

Comment (ndr-brt) [reply .github/workflows/release.yml:40]: Either having this workflow to run when a PR is closed on a `release` branch or making it manual as for the bugfix version. 
At the end of the release we could merge back to main if ref_name starts with `release` and snapshot version can be bumped.
The more the `release` and `bugfix` process are similar, the easier will be to maintain them

Comment (rafaelmag110) [reply .github/workflows/release.yml:40]: The workflow is being triggered when a PR is closed on a `release` branch. If we really want to avoid the workflow to be evalutated on every closed PR to main, we can make it manual only.

File: .github/actions/publish-tag-and-release/action.yml
Comment (rafaelmag110) [reply .github/actions/publish-tag-and-release/action.yml]: [3ea619b](https://github.com/eclipse-tractusx/tractusx-edc/pull/1543/commits/3ea619b7c4e94c36c9efbb977336e6dbd521b639)

File: .github/actions/publish-tag-and-release/action.yml
Comment (rafaelmag110) [reply .github/actions/publish-tag-and-release/action.yml]: [3ea619b](https://github.com/eclipse-tractusx/tractusx-edc/pull/1543/commits/3ea619b7c4e94c36c9efbb977336e6dbd521b639)

File: .github/actions/publish-tag-and-release/action.yml
Comment (rafaelmag110) [reply .github/actions/publish-tag-and-release/action.yml]: [3ea619b](https://github.com/eclipse-tractusx/tractusx-edc/pull/1543/commits/3ea619b7c4e94c36c9efbb977336e6dbd521b639)

File: .github/workflows/draft-release.yaml
Comment (rafaelmag110) [reply .github/workflows/draft-release.yaml]: I refactored the validation jobs in [788baa1](https://github.com/eclipse-tractusx/tractusx-edc/pull/1543/commits/788baa13d5600130017551642181721f394b7682) since I also agree those could be simpler.

On a personal level, I prefer the conditionals on the jobs or actions. The workflow execution graph is calculated before it is sent to the runner and false conditionals are skipped. When having everything in a shell script you have to be more careful what behaviors should lead to job failure and explicitly have it fail.

I'm fine either way. Let me know if the refactor is ok.

Comment (paullatzelsperger) [reply .github/workflows/draft-release.yaml]: We've seen this numerous times, where we want to replicate the runner's behaviour and debug or test a workflow locally, e.g. after a refactoring or a maintenance change. If everything is a job with conditional execution, there is zero chance of doing that, and ACT only gets you so far.

Let's get this merged as-is so as not to block the PR. 

Generally speaking, I would be very cautious when adding GH actions, because we are effectively introducing undebuggable foreign code, that you have no control over.

**Review by paullatzelsperger — APPROVED** [review]

**Pull request merged**
